From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: NONPLAYT <nonplay@bxteam.org>
Date: Wed, 19 Mar 2025 23:24:32 +0300
Subject: [PATCH] Pufferfish: Optimize mob spawning

Original license: GPL v3
Original project: https://github.com/pufferfish-gg/Pufferfish

This patch reduces the main-thread impact of mob spawning by moving spawning work to other threads

diff --git a/net/minecraft/server/MinecraftServer.java b/net/minecraft/server/MinecraftServer.java
index a6aad24383b23f7994400468a4783df9972217bc..3d790d7cda79e19a2f3f891bc408196d3d3624c5 100644
--- a/net/minecraft/server/MinecraftServer.java
+++ b/net/minecraft/server/MinecraftServer.java
@@ -340,6 +340,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
     public boolean lagging = false; // Purpur - Lagging threshold
     protected boolean upnp = false; // Purpur - UPnP Port Forwarding
     public final org.bxteam.divinemc.util.tps.TPSCalculator tpsCalculator = new org.bxteam.divinemc.util.tps.TPSCalculator(); // DivineMC - Lag compensation
+    public gg.pufferfish.pufferfish.util.AsyncExecutor mobSpawnExecutor = new gg.pufferfish.pufferfish.util.AsyncExecutor("Mob Spawning"); // DivineMC - Pufferfish: Optimize mob spawning
     // Paper start - improve tick loop
     public final ca.spottedleaf.moonrise.common.time.TickData tickTimes1s  = new ca.spottedleaf.moonrise.common.time.TickData(java.util.concurrent.TimeUnit.SECONDS.toNanos(1L));
     public final ca.spottedleaf.moonrise.common.time.TickData tickTimes5s  = new ca.spottedleaf.moonrise.common.time.TickData(java.util.concurrent.TimeUnit.SECONDS.toNanos(5L));
diff --git a/net/minecraft/server/dedicated/DedicatedServer.java b/net/minecraft/server/dedicated/DedicatedServer.java
index 060c4c2c0ba29c9f8c26b4ea7703950893160a3e..6ee80440be2b39e343a260e3f00cf59c46496eae 100644
--- a/net/minecraft/server/dedicated/DedicatedServer.java
+++ b/net/minecraft/server/dedicated/DedicatedServer.java
@@ -479,7 +479,7 @@ public class DedicatedServer extends MinecraftServer implements ServerInterface
 
         org.purpurmc.purpur.task.BossBarTask.startAll(); // Purpur - Implement TPSBar
         if (org.purpurmc.purpur.PurpurConfig.beeCountPayload) org.purpurmc.purpur.task.BeehiveTask.instance().register(); // Purpur - Give bee counts in beehives to Purpur clients
-
+        if (org.bxteam.divinemc.config.DivineConfig.AsyncCategory.enableAsyncSpawning) mobSpawnExecutor.start(); // DivineMC - Pufferfish: Optimize mob spawning
         this.saveEverything(false, true, true);
         this.notificationManager().serverStarted();
         return true;
diff --git a/net/minecraft/server/level/ServerChunkCache.java b/net/minecraft/server/level/ServerChunkCache.java
index 01d9852953806f7132ab6d1f7a532d86419e7904..e315c11117b77d81eed442d8ec6cc856fd052336 100644
--- a/net/minecraft/server/level/ServerChunkCache.java
+++ b/net/minecraft/server/level/ServerChunkCache.java
@@ -75,6 +75,10 @@ public class ServerChunkCache extends ChunkSource implements ca.spottedleaf.moon
     private final Set<ChunkHolder> chunkHoldersToBroadcast = new ReferenceOpenHashSet<>();
     @VisibleForDebug
     private NaturalSpawner.@Nullable SpawnState lastSpawnState;
+    // DivineMC start - Pufferfish: Optimize mob spawning
+    public boolean firstRunSpawnCounts = true;
+    public final java.util.concurrent.atomic.AtomicBoolean spawnCountsReady = new java.util.concurrent.atomic.AtomicBoolean(false);
+    // DivineMC end - Pufferfish: Optimize mob spawning
     // Paper start
     public final ca.spottedleaf.concurrentutil.map.concurrent.longs.ConcurrentChainedLong2ReferenceHashTable<net.minecraft.world.level.chunk.LevelChunk> fullChunks = new ca.spottedleaf.concurrentutil.map.concurrent.longs.ConcurrentChainedLong2ReferenceHashTable<>();
     public int getFullChunksCount() {
@@ -516,6 +520,47 @@ public class ServerChunkCache extends ChunkSource implements ca.spottedleaf.moon
             this.broadcastChangedChunks(profiler);
             profiler.pop();
         }
+
+        // DivineMC start - Pufferfish: Optimize mob spawning
+        if (org.bxteam.divinemc.config.DivineConfig.AsyncCategory.enableAsyncSpawning) {
+            for (ServerPlayer player : this.level.players()) {
+                for (int ii = 0; ii < ServerPlayer.MOBCATEGORY_TOTAL_ENUMS; ii++) {
+                    player.mobCounts[ii] = 0;
+
+                    int newBackoff = Math.max(0, player.mobBackoffCounts[ii] - 1);
+                    player.mobBackoffCounts[ii] = newBackoff;
+                }
+            }
+
+            if (firstRunSpawnCounts) {
+                firstRunSpawnCounts = false;
+                spawnCountsReady.set(true);
+            }
+
+            if (spawnCountsReady.getAndSet(false)) {
+                net.minecraft.server.MinecraftServer.getServer().mobSpawnExecutor.submit(() -> {
+                    int mapped = distanceManager.getNaturalSpawnChunkCount();
+                    ca.spottedleaf.moonrise.common.list.IteratorSafeOrderedReferenceSet.Iterator<Entity> objectiterator = level.entityTickList.entities.iterator(ca.spottedleaf.moonrise.common.list.IteratorSafeOrderedReferenceSet.ITERATOR_FLAG_SEE_ADDITIONS);
+
+                    try {
+                        gg.pufferfish.pufferfish.util.IterableWrapper<Entity> wrappedIterator = new gg.pufferfish.pufferfish.util.IterableWrapper<>(objectiterator);
+                        LocalMobCapCalculator mobCapCalculator = !level.paperConfig().entities.spawning.perPlayerMobSpawns ? new LocalMobCapCalculator(chunkMap) : null;
+
+                        lastSpawnState = NaturalSpawner.createState(
+                            mapped,
+                            wrappedIterator,
+                            ServerChunkCache.this::getFullChunk,
+                            mobCapCalculator,
+                            level.paperConfig().entities.spawning.perPlayerMobSpawns
+                        );
+                    } finally {
+                        objectiterator.finishedIterating();
+                    }
+                    spawnCountsReady.set(true);
+                });
+            }
+        }
+        // DivineMC end - Pufferfish: Optimize mob spawning
     }
 
     private void broadcastChangedChunks(final ProfilerFiller profiler) {
@@ -536,28 +581,31 @@ public class ServerChunkCache extends ChunkSource implements ca.spottedleaf.moon
         profiler.push("naturalSpawnCount");
         int chunkCount = this.distanceManager.getNaturalSpawnChunkCount();
         // Paper start - Optional per player mob spawns
-        NaturalSpawner.SpawnState spawnCookie;
+        // DivineMC start - Pufferfish: Optimize mob spawning
         if ((this.spawnFriendlies || this.spawnEnemies) && this.level.paperConfig().entities.spawning.perPlayerMobSpawns) { // don't count mobs when animals and monsters are disabled
-            // re-set mob counts
-            for (ServerPlayer player : this.level.players()) {
-                // Paper start - per player mob spawning backoff
-                for (int j = 0; j < ServerPlayer.MOBCATEGORY_TOTAL_ENUMS; j++) {
-                    player.mobCounts[j] = 0;
-
-                    int newBackoff = player.mobBackoffCounts[j] - 1; // TODO make configurable bleed // TODO use nonlinear algorithm?
-                    if (newBackoff < 0) {
-                        newBackoff = 0;
+            if (!org.bxteam.divinemc.config.DivineConfig.AsyncCategory.enableAsyncSpawning) {
+                // re-set mob counts
+                for (ServerPlayer player : this.level.players()) {
+                    // Paper start - per player mob spawning backoff
+                    for (int j = 0; j < ServerPlayer.MOBCATEGORY_TOTAL_ENUMS; j++) {
+                        player.mobCounts[j] = 0;
+
+                        int newBackoff = player.mobBackoffCounts[j] - 1; // TODO make configurable bleed // TODO use nonlinear algorithm?
+                        if (newBackoff < 0) {
+                            newBackoff = 0;
+                        }
+                        player.mobBackoffCounts[j] = newBackoff;
                     }
-                    player.mobBackoffCounts[j] = newBackoff;
+                    // Paper end - per player mob spawning backoff
                 }
-                // Paper end - per player mob spawning backoff
+                this.lastSpawnState = NaturalSpawner.createState(chunkCount, this.level.getAllEntities(), this::getFullChunk, null, true);
             }
-            spawnCookie = NaturalSpawner.createState(chunkCount, this.level.getAllEntities(), this::getFullChunk, null, true);
         } else {
-            spawnCookie = NaturalSpawner.createState(chunkCount, this.level.getAllEntities(), this::getFullChunk, !this.level.paperConfig().entities.spawning.perPlayerMobSpawns ? new LocalMobCapCalculator(this.chunkMap) : null, false);
+            this.lastSpawnState = NaturalSpawner.createState(chunkCount, this.level.getAllEntities(), this::getFullChunk, !this.level.paperConfig().entities.spawning.perPlayerMobSpawns ? new LocalMobCapCalculator(this.chunkMap) : null, false);
+            spawnCountsReady.set(true);
         }
+        // DivineMC end - Pufferfish: Optimize mob spawning
         // Paper end - Optional per player mob spawns
-        this.lastSpawnState = spawnCookie;
         boolean doMobSpawning = this.level.getGameRules().get(GameRules.SPAWN_MOBS) && !this.level.players().isEmpty(); // CraftBukkit
         int tickSpeed = this.level.getGameRules().get(GameRules.RANDOM_TICK_SPEED);
         List<MobCategory> spawningCategories;
@@ -571,7 +619,7 @@ public class ServerChunkCache extends ChunkSource implements ca.spottedleaf.moon
             }
             // Paper end - PlayerNaturallySpawnCreaturesEvent
             boolean spawnPersistent = this.level.ticksPerSpawnCategory.getLong(org.bukkit.entity.SpawnCategory.ANIMAL) != 0L && this.level.getGameTime() % this.level.ticksPerSpawnCategory.getLong(org.bukkit.entity.SpawnCategory.ANIMAL) == 0L; // CraftBukkit
-            spawningCategories = NaturalSpawner.getFilteredSpawningCategories(spawnCookie, this.spawnFriendlies, this.spawnEnemies, spawnPersistent, this.level); // CraftBukkit
+            spawningCategories = NaturalSpawner.getFilteredSpawningCategories(this.lastSpawnState, this.spawnFriendlies, this.spawnEnemies, spawnPersistent, this.level); // CraftBukkit // DivineMC - Pufferfish: Optimize mob spawning
         } else {
             spawningCategories = List.of();
         }
@@ -591,7 +639,7 @@ public class ServerChunkCache extends ChunkSource implements ca.spottedleaf.moon
             profiler.popPush("tickSpawningChunks");
 
             for (LevelChunk chunk : spawningChunks) {
-                this.tickSpawningChunk(chunk, timeDiff, spawningCategories, spawnCookie);
+                this.tickSpawningChunk(chunk, timeDiff, spawningCategories, this.lastSpawnState); // DivineMC - Pufferfish: Optimize mob spawning
             }
         } finally {
             spawningChunks.clear();
@@ -616,11 +664,11 @@ public class ServerChunkCache extends ChunkSource implements ca.spottedleaf.moon
             this.level.tickThunder(chunk);
         }
 
-        if (!spawningCategories.isEmpty()) {
-            if (this.level.getWorldBorder().isWithinBounds(chunkPos)) { // Paper - rewrite chunk system
-                NaturalSpawner.spawnForChunk(this.level, chunk, spawnCookie, spawningCategories);
-            }
+        // DivineMC start - Pufferfish: Optimize mob spawning
+        if (!spawningCategories.isEmpty() && this.level.getWorldBorder().isWithinBounds(chunkPos) && (!org.bxteam.divinemc.config.DivineConfig.AsyncCategory.enableAsyncSpawning || spawnCountsReady.get())) { // Paper - rewrite chunk system
+            NaturalSpawner.spawnForChunk(this.level, chunk, spawnCookie, spawningCategories);
         }
+        // DivineMC end - Pufferfish: Optimize mob spawning
     }
 
     private void getFullChunk(final long chunkKey, final Consumer<LevelChunk> output) {
